### PR TITLE
Add PCAM dataset to prototype area

### PR DESCRIPTION
**Pull request by NicolasHug**

### CODE REVIEW

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (NicolasHug) [inline torchvision/prototype/datasets/_builtin/pcam.py:43]: The .h5 files store decoded numpy arrays, so I assume `RAW` is the right one here?

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (NicolasHug) [inline torchvision/prototype/datasets/_builtin/pcam.py:34]: Note: using the file_path (`_`) works just as well here. @pmeier are there any subtleties that would make us choose one over the other?

File: test/builtin_dataset_mocks.py
Comment (jdsgomes) [reply test/builtin_dataset_mocks.py:1341]: I think if you want to do everything in memory this pattern would work:

```python
images_io = io.BytesIO()
with h5py.File(images_io, 'w') as f:
    f["x"] = np.random.randint(0, 256, size=(num_images, 10, 10, 3), dtype=np.uint8)
with open(compressed_file_name, "wb") as compressed_file:
    compressed_file.write(gzip.compress(images_io.read()))
```

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/pcam.py:26]: Nit: I would reserve line comments for linter directives.

```suggestion
        # Note: this key thing might be very specific to the PCAM dataset
        key: Optional[str] = None,
```

This applies to all occurrences of this.

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/pcam.py:22]: Maybe we can call it 

```suggestion
class PCAMH5Reader(IterDataPipe[Tuple[str, io.IOBase]]):
```

and only make it more general if another dataset also needs this?

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/pcam.py:32]: What is `flake8` complaining here? In any case can we make this more specific like `# noqa: F401`.

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (pmeier) [reply torchvision/prototype/datasets/_builtin/pcam.py:34]: 1. The file is already open so there is no need to open it again. I'm guessing this works on Linux and Mac, but could maybe bite us on Windows.
2. For this dataset the files are present on disk directly. If they live inside an archive, the path is just an indicator and not a real path, like `/foo/bar.zip/baz.h5`. In this case we can only use the handle.

For these reasons I would always go with the handle directly unless there is a reason not to.

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/pcam.py:50]: This is kinda hard to parse. Maybe we can have small dataclass or namedtuple to carry the information?

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/pcam.py:101]: Internally, the custom `Feature`'s call 

https://github.com/pytorch/vision/blob/11d903ec6b8401d072c48d42ab0c3a6e3c8f62c0/torchvision/prototype/features/_feature.py#L76-L78

on the input. Thus, we don't need to convert manually here

```suggestion
            "image": features.Image(image),
            "label": Label(target),
```

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (pmeier) [inline torchvision/prototype/datasets/_builtin/pcam.py:122]: Nit: saw this a couple of times in your PRs. Any advantage of returning directly?

```suggestion
        return Mapper(dp, self._collate_and_decode)
```

**Review by pmeier — APPROVED** [review]
Only minor comments inline. Thanks a lot @NicolasHug! LGTM when CI is green.

File: test/builtin_dataset_mocks.py
Comment (NicolasHug) [reply test/builtin_dataset_mocks.py:1341]: Thanks @jdsgomes !
For some reason I had to use `getbuffer()` instead of `read()`.

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (NicolasHug) [reply torchvision/prototype/datasets/_builtin/pcam.py:123]: It's just more consistent with the rest of the code which uses

`dp = ...`

a lot and it makes it obvious that what it returned is still a "dp" for non-expert readers. But I don't feel strongly about this

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (NicolasHug) [reply torchvision/prototype/datasets/_builtin/pcam.py:26]: Do you feel strongly about this? I like line comments because it's very clear what they apply to.
In some occasions, I have a hard time figuring out whether a comment applies to a single line below, or to a bigger block

**Review by jdsgomes — APPROVED** [review]
LGTM, there are just some errors related to imports, once those are sorted the PR is ready to be merged

File: torchvision/prototype/datasets/_builtin/pcam.py
Comment (pmeier) [reply torchvision/prototype/datasets/_builtin/pcam.py:27]: No strong opinion.

**Pull request merged**
